gst/nnstreamer/include/nnstreamer_plugin_api_filter.h

```diff
@@ -615,6 +615,84 @@ extern void
+  unsigned int num_labels;      /* The number of label lists, the label is where framework receive the class to train the model, num_labels indicates how many labels there are. */
+  unsigned int num_train_samples;       /* The number of train sample used to train the model. */
+  unsigned int num_valid_samples;       /* The number of valid sample used to train the model. */
+} GstTensorTrainerProperties;
+
+
+/**
+ * @brief tensor_trainer Subplugin definition
+ *
+ * Common callback parameters:
+ * prop Trainer properties. Read Only.
+ * private_data Subplugin's private data. Set this (*private_data = XXX) if you want to change trainer->private_data.
+ */
+typedef struct _GstTensorTrainerFramework
+{
+  /* The name of the neural network, searchable by 'framework' property */
+  char *name;
+
+  /**
+   * tensor_trainer call this to create the model
+   * @param[in] prop read-only property values
+   * @param[in/out] private_data, a subplugin may save its internal private data here.
+   * @return 0 if ok. < 0 if error.
+   */
+  int (*create) (const GstTensorTrainerProperties * prop, void **private_data);
+
+  /**
+   * tensor_trainer call this to destroy the model, Set NULL after that.
+   * @param[in] prop read-only property values
+   * @param[in/out] private_data, a subplugin may save its internal private data here.
+   * @return 0 if ok. < 0 if error.
+   */
+  int (*destroy) (const GstTensorTrainerProperties * prop, void **private_data);
+
+  /**
+   * tensor_trainer call this to train the model
+   * @param[in] prop read-only property values
+   * @param[in/out] private_data, a subplugin may save its internal private data here.
+   * @return 0 if ok. < 0 if error.
+   */
+  int (*train) (const GstTensorTrainerProperties * prop, void **private_data);
+
+  /**
+   * tensor_trainer call this to send tensor data to subplugin
+   * @param[in] prop read-only property values
+   * @param[in/out] private_data, a subplugin may save its internal private data here.
+   * @param[in] input The array of input tensors. Allocated and filled by tensor_trainer
+   * @param[out] output The array of output tensors. Allocated by tensor_trainer, set to NULL if not used.
+   */
+  int (*invoke_NN) (const GstTensorTrainerProperties * prop,
+      void **private_data, const GstTensorMemory * input,
+      GstTensorMemory * output);
+
+} GstTensorTrainerFramework;
+
+/* extern functions for subplugin management, exist in tensor_trainer.c */
+/**
+ * @brief Trainer's sub-plugin should call this function to register itself.
+ * @param[in] sub_plugin tensor_trainer sub-plugin to be registered.
+ * @return TRUE if registered. FALSE is failed or duplicated.
+ *
+ * @note Do not change the subplugins callbacks after probing the filter.
+ */
+extern int
+nnstreamer_trainer_probe (GstTensorTrainerFramework * sub_plugin);
```

Comment:
constify sub_plugin param?

Reply:
thank you.
I will fix it like tensor_filter.

/* extern functions for subplugin management, exist in tensor_filter.c */
/**
 * @brief Filter's sub-plugin should call this function to register itself.
 * @param[in] tfsp Tensor-Filter Sub-Plugin to be registered.
 * @return TRUE if registered. FALSE is failed or duplicated.
 *
 * @note Do not change the subplugins callbacks after probing the filter.
 */
extern int
nnstreamer_filter_probe (GstTensorFilterFramework * tfsp);